okio-files/src/macosX64Main/kotlin/okio/FileSink.kt

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2020 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio
+
+import kotlinx.cinterop.ByteVar
+import kotlinx.cinterop.CPointer
+import kotlinx.cinterop.allocArray
+import kotlinx.cinterop.nativeHeap
+import platform.posix.FILE
+import platform.posix.errno
+import platform.posix.fclose
+import platform.posix.fflush
+import platform.posix.free
+import platform.posix.fwrite
+
+/** Writes bytes to a file as a sink. */
+internal class FileSink(
+  private val file: CPointer<FILE>
+) : Sink {
+  private var nativeBuffer = nativeHeap.allocArray<ByteVar>(segmentSize)
+  private var closed = false
```

Comment:
General question: should we be concerned about concurrency at this level, i.e. should these be volatile, atomic, or should the methods use locks etc.?

Reply:
Sources and Sinks are not thread-safe, like an InputStream or OutputStream. You have to guard things above them.

Reply:
Fun fact: InputStream and OutputStream use synchronization, but it's the bad ill-advised synchronization of StringBuffer and Vector of 1990s Java. 

Either way, sources and sinks and buffers are not safe for concurrent use.